File: CartMinder/app/src/main/java/id/ac/ui/cs/mobileprogramming/kevinlh/cartminder/repository/CartRepository.java
package id.ac.ui.cs.mobileprogramming.kevinlh.cartminder.repository;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import java.lang.ref.WeakReference;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import id.ac.ui.cs.mobileprogramming.kevinlh.cartminder.dao.CartDAO;
import id.ac.ui.cs.mobileprogramming.kevinlh.cartminder.database.Database;
import id.ac.ui.cs.mobileprogramming.kevinlh.cartminder.model.Cart;
import id.ac.ui.cs.mobileprogramming.kevinlh.cartminder.model.Item;

public class CartRepository {
    private ExecutorService daoExecutor;
    private CartDAO cartDAO;
    private LiveData<List<Cart>> carts;

    public CartRepository(Application application) {
        Database database = Database.getInstance(application);
        daoExecutor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS,
                new LinkedBlockingQueue<Runnable>());
        cartDAO = database.cartDAO();
        carts = cartDAO.getCartsOrdered();
    }

    public long insert(Cart cart) {
        try {
            InsertCartAsyncTask task = new InsertCartAsyncTask(this);
            return task.execute(cart).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            System.out.println("[error]" + e.getMessage());
            return -1;
        }
    }

    public void update(Cart cart) {
        daoExecutor.execute(() -> {
            cartDAO.update(cart);
        });
    }

    public void delete(Cart cart) {
        daoExecutor.execute(() -> {
            cartDAO.delete(cart);
        });
    }

    public LiveData<List<Cart>> getCarts() {
        return carts;
    }

    private static class InsertCartAsyncTask extends AsyncTask<Cart, Void, Long> {
        private WeakReference<CartRepository> weakReference;

        InsertCartAsyncTask(CartRepository cartRepository) {
            weakReference = new WeakReference<CartRepository>(cartRepository);
        }

        @Override
        protected Long doInBackground(Cart... carts) {
            CartRepository reference = weakReference.get();
            if (reference == null) return null;
            return reference.cartDAO.insert(carts[0]);
        }
    }
}

File: CartMinder2/app/src/main/java/id/ac/ui/cs/mobileprogramming/kevinlh/cartminder2/database/repository/CartRepository.kt
package id.ac.ui.cs.mobileprogramming.kevinlh.cartminder2.database.repository

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.asLiveData
import id.ac.ui.cs.mobileprogramming.kevinlh.cartminder2.database.CartMinderDatabase
import id.ac.ui.cs.mobileprogramming.kevinlh.cartminder2.database.dao.CartDao
import id.ac.ui.cs.mobileprogramming.kevinlh.cartminder2.database.dao.ItemDao
import id.ac.ui.cs.mobileprogramming.kevinlh.cartminder2.model.Cart
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class CartRepository(application: Application) {
    private val cartDao: CartDao
    private val itemDao: ItemDao

    init {
        val database = CartMinderDatabase.getInstance(application)
        cartDao = database.cartDao
        itemDao = database.itemDao
    }

    suspend fun findById(cartId: Long): Cart = cartDao.findById(cartId)

    suspend fun insertCart(cart: Cart): Long = cartDao.insert(cart)

    suspend fun getTotalPrice(cart: Cart): Int {
        var total = 0
        val items = itemDao.findByCartId(cart.id)
        for (i in items) {
            total += i.price
        }
        return total
    }

    fun getLiveCarts(): LiveData<List<Cart>> = cartDao.getCarts().asLiveData()

    fun updateCart(cart: Cart) = CoroutineScope(Dispatchers.IO).launch { cartDao.update(cart) }

    fun deleteCart(cart: Cart) = CoroutineScope(Dispatchers.IO).launch { cartDao.delete(cart) }

    fun deleteAll() {
        CoroutineScope(Dispatchers.IO).launch {
            cartDao.findAll().forEach {
                deleteCart(it)
            }
        }
    }

    companion object {
        private var instance: CartRepository? = null

        fun getInstance(application: Application): CartRepository {
            return instance ?: CartRepository(application)
        }
    }
}